File: server/src/controllers/request.js
const Request = require("../models/request");
const Appointment = require("../models/request");
const BloodType = require("../models/bloodType");

const { validationResult } = require("express-validator");
const User = require("../models/user");
const { model } = require("mongoose");

exports.getAllRequest = async (req, res, next) => {
  try {
    const page = req.query.page * 1 || 1;
    const limit = req.query.limit * 1 || 10;

    const result = await Request.paginate(
      { isDeleted: false },
      {
        populate: {
          path: "bloodType createdBy updatedBy donors",
          populate: { path: "donors", model: "User" },
        },
        page,
        limit,
        sort: "-createdAt",
      }
    );
    res.status(200).json({
      status: "success",
      result,
    });
  } catch (error) {
    console.log(error);
  }
};

exports.getRequest = async (req, res, next) => {
  try {
    const request = await Request.findOne({
      _id: req.params.id,
      isDeleted: false,
    }).populate("bloodType donors createdBy updatedBy");

    if (!request) {
      return res.status(404).json({
        status: "error",
        message: "Request with this ID does not exist",
      });
    }
    if (request.isDeleted) {
      return res.status(404).json({
        status: "error",
        message: "Request with this ID does not exist",
      });
    }
    res.status(200).json({
      status: "success",
      request,
    });
  } catch (error) {
    console.log(error);
  }
};

exports.createRequest = async (req, res, next) => {
  try {
    console.log(req.body.bloodType + "**********************");
    const userBloodType = await BloodType.findOne({
      bloodTypeName: req.body.bloodType,
    });

    let request = await Request.create({
      ...req.body,
      bloodType: userBloodType._id,
      createdBy: req.user._id,
      updatedBy: req.user._id,
      donors: [req.user._id],
    });
    request = await Request.findOne({
      _id: request._id,
      isDeleted: false,
    }).populate("bloodType donors created");

    res.status(201).json({
      status: "success",
      request,
    });
  } catch (error) {
    console.log(error);
  }
};

exports.updateRequest = async (req, res, next) => {
  try {

    const userBloodType = await BloodType.findOne({
      bloodTypeName: req.body.bloodType,
    });

    let request = await Request.findByIdAndUpdate(
      req.params.id,
      { ...req.body, bloodType: userBloodType },
      {
        new: true,
      }
    );
    request = await Request.findOne({
      _id: request._id,
      isDeleted: false,
    }).populate("bloodType donors created");

    if (!request) {
      return res.status(404).json({
        status: "error",
        message: "Request with this ID does not exist",
      });
    }
    res.status(200).json({
      status: "success",
      request,
    });
  } catch (error) {
    console.log(error);
  }
};

exports.deleteRequest = async (req, res, next) => {
  try {
    const request = await Request.findByIdAndUpdate(req.params.id, {
      isDeleted: true,
    });

    if (!request) {
      return res.status(404).json({
        status: "error",
        message: "Request with this ID does not exist",
      });
    }
    res.status(200).json({
      status: "success",
      request: null,
    });
  } catch (error) {
    console.log(error);
  }
};

exports.acceptRequest = async (req, res, next) => {
  try {

    let request = await Request.findOne({
      _id: req.params.id,
      isDeleted: false,
    });

    if (!request) {
      return res.status(404).json({
        status: "error",
        message: "Request with this ID does not exist",
      });
    }

    request = await Request.updateOne(
      {
        _id: req.params.id,
      },
      { $inc: { totalDonations: 1 }, $push: { donors: req.user._id } }
    ).exec();

    request = await Request.findOne({
      _id: req.params.id,
      isDeleted: false,
    }).populate({
      path: "bloodType donors createdBy",
      populate: { path: "donors", model: "User" },
    });

    let appointment = await Appointment.create({
      userId: req.user._id,
      startDate: new Date().toISOString(),
      endDate: new Date(
        new Date().getTime() + 15 * 24 * 60 * 60 * 1000
      ).toISOString(),
      appointmentDescription: `Appointment using Request ${request._id}: Reason: ${request.reason}`,
    });

    res.status(200).json({
      status: "success",
      request,
    });
  } catch (error) {
    console.log(error);
  }
};

File: client/lib/src/request/models/request.dart
import 'package:eshiblood/src/request/model/donor.dart';

class Request {
  Request({
    this.id,
    required this.unitsNeeded,
    this.totalDonations,
    required this.reason,
    this.status,
    required this.bloodType,
    this.createdBy,
    this.UpdatedBy,
    this.donors,
  });
  String? id;
  dynamic unitsNeeded;
  dynamic totalDonations;
  String reason;
  String? status;
  String? bloodType;
  String? createdBy;
  String? UpdatedBy;
  List<Donor>? donors;

  factory Request.fromJson(Map<String, dynamic> json) {
    String? createdBy;
    String? updatedBy;
    try {
      createdBy = json["createdBy"]["_id"];
    } catch (e) {
      createdBy = json["createdBy"];
    }

    try {
      updatedBy = json["updatedBy"]["_id"];
    } catch (e) {
      updatedBy = json["updatedBy"];
    }

    return Request(
      id: json["_id"],
      unitsNeeded: json["unitsNeeded"] ?? 0,
      totalDonations: json["totalDonations"] ?? 0,
      reason: json["reason"],
      status: json["status"],
      bloodType: json["bloodType"]["bloodTypeName"],
      createdBy: createdBy,
      UpdatedBy: updatedBy,
      donors: List<Donor>.generate(json["donors"].length,
          (index) => Donor.fromJson(json["donors"][index])),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      "_id": id,
      "unitsNeeded": unitsNeeded,
      "totalDonations": totalDonations,
      "reason": reason,
      "status": status,
      "bloodType": bloodType,
      "createdBy": createdBy,
      "UpdatedBy": UpdatedBy,
      "donors": donors,
    };
  }

  @override
  String toString() {
    return """{\n
        _id: $id, \n
        unitsNeeded: $unitsNeeded,\n
        totalDonations: $totalDonations,\n
        reason: $reason,\n
        status: $status,\n
        bloodType: $bloodType,\n
        createdBy: $createdBy,\n
        UpdatedBy: $UpdatedBy,\n
        donors: $donors,\n
    }""";
  }
}
